### PR TITLE
feat(timeline): create package

### PR DESCRIPTION
**Pull request by PixeledCode**

### CODE REVIEW

File: packages/paste-core/components/timeline/src/TimelineItemGroup.tsx
Comment (serifluous) [inline packages/paste-core/components/timeline/src/TimelineItemGroup.tsx:52]: question: @twilio-labs/design-systems-eng thoughts on adding a marginBottom prop to DetailText to clean up some of the extra spacing we have to do here?

I've seen 8px bottom spacing show up a few times with Detail Text compositions now (see https://paste.twilio.design/components/page-header#page-header-on-a-wizard-page) that I think it could make sense.

cc @roxanagomez

File: packages/paste-core/components/timeline/src/Timeline.tsx
Comment (krisantrobus) [inline packages/paste-core/components/timeline/src/Timeline.tsx:9]: praise: love the type safety here

File: packages/paste-core/components/timeline/src/index.tsx
Comment (krisantrobus) [inline packages/paste-core/components/timeline/src/index.tsx:5]: nit: I don't think Orientation will need to be exported to users as it will still use the type values for props
```suggestion
export type { TimelineProps, TimelineItemProps, TimelineItemIconProps } from "./types";
```

File: packages/paste-core/components/timeline/src/types.ts
Comment (krisantrobus) [inline packages/paste-core/components/timeline/src/types.ts:16]: Just a comment for other reviewers, I'm fine with leaving this in if we are implementing at some point in the future

File: packages/paste-core/components/timeline/src/TimelineItemIcon.tsx
Comment (krisantrobus) [inline packages/paste-core/components/timeline/src/TimelineItemIcon.tsx:16]: has this typing been fixed? Any reason we have to ignore this?

File: packages/paste-core/components/timeline/src/TimelineItemIcon.tsx
Comment (krisantrobus) [inline packages/paste-core/components/timeline/src/TimelineItemIcon.tsx:28]: can you use the space tokens here
```suggestion
            size="sizeBase20"
```

File: packages/paste-core/components/timeline/src/TimelineItemGroup.tsx
Comment (krisantrobus) [inline packages/paste-core/components/timeline/src/TimelineItemGroup.tsx:45]:
```suggestion
        width={orientation === "horizontal" ? "initial" : "sizeBase50"}
```

File: packages/paste-core/components/timeline/src/TimelineItem.tsx
Comment (krisantrobus) [inline packages/paste-core/components/timeline/src/TimelineItem.tsx:28]: is this needed? I'm inspecting the grouped an non grouped versions for element="TIMELINE_ITEM_ICON_WRAPPER " and in both the ::after has display block?

File: packages/paste-core/components/timeline/src/TimelineItem.tsx
Comment (krisantrobus) [inline packages/paste-core/components/timeline/src/TimelineItem.tsx:52]:
```suggestion
            width={orientation === "horizontal" ? "initial" : "sizeBase50"}
```

**Review by roxanagomez — COMMENTED** [review]
TODO: remove the horizontal variant from the stories.
Screenshot 2024-09-05 at 3 31 45 PM

File: packages/paste-core/components/timeline/src/TimelineItem.tsx
Comment (PixeledCode) [reply packages/paste-core/components/timeline/src/TimelineItem.tsx:28]: yeah, removing it causes it to look like this:
image

File: packages/paste-core/components/timeline/src/TimelineItemGroup.tsx
Comment (nkrantz) [reply packages/paste-core/components/timeline/src/TimelineItemGroup.tsx]: my only consideration is that it's a little weird to have the default marginTop be not 0 and have to reset it and the default marginBottom be 0 and have to add it. otherwise into it

File: packages/paste-core/components/timeline/src/types.ts
Comment (nkrantz) [reply packages/paste-core/components/timeline/src/types.ts]: hmm I'd rather remove all the commented stuff and keep it in a branch for the future

File: .changeset/famous-experts-speak.md
Comment (nkrantz) [inline .changeset/famous-experts-speak.md:3]: 👍

File: packages/paste-core/components/timeline/__tests__/index.spec.tsx
Comment (nkrantz) [inline packages/paste-core/components/timeline/__tests__/index.spec.tsx:71]: suggestion(optional): you could move this line outside of the `it` blocks so that you only have to render it once as every block in this describe is using the same line

File: packages/paste-core/components/timeline/__tests__/index.spec.tsx
Comment (nkrantz) [inline packages/paste-core/components/timeline/__tests__/index.spec.tsx:73]: suggestion(optional): another way to test here could be to make sure the <ol> and <li>s are in the document

File: packages/paste-core/components/timeline/package.json
Comment (nkrantz) [inline packages/paste-core/components/timeline/package.json:6]:
```suggestion
  "description": "A Timeline is a visual representation of events displayed in chronological order.",
```

File: packages/paste-core/components/timeline/src/Timeline.tsx
Comment (nkrantz) [inline packages/paste-core/components/timeline/src/Timeline.tsx:30]:
```suggestion
      minHeight: "sizeBase80",
```

File: packages/paste-core/components/timeline/src/Timeline.tsx
Comment (nkrantz) [reply packages/paste-core/components/timeline/src/Timeline.tsx:9]: +1!! but we probably don't need VariantStyles anymore if we're not doing horizontal right? same with the keys in ItemSeparatorStyles

File: packages/paste-core/components/timeline/src/Timeline.tsx
Comment (nkrantz) [inline packages/paste-core/components/timeline/src/Timeline.tsx:58]: question: what does setting the ol as display=flex do here?

File: packages/paste-core/components/timeline/src/Timeline.tsx
Comment (nkrantz) [inline packages/paste-core/components/timeline/src/Timeline.tsx:60]:
```suggestion
        ...VariantStyles,
        ...ItemSeparatortyles,
```
once we remove horizontal styles

File: packages/paste-core/components/timeline/src/Timeline.tsx
Comment (nkrantz) [inline packages/paste-core/components/timeline/src/Timeline.tsx:21]:
```suggestion
const ItemSeparatorStyles: {
```

File: packages/paste-core/components/timeline/src/TimelineContext.tsx
Comment (nkrantz) [inline packages/paste-core/components/timeline/src/TimelineContext.tsx:1]: suggestion: remove context for now as it's only being used for orientation

File: packages/paste-core/components/timeline/src/types.ts
Comment (nkrantz) [inline packages/paste-core/components/timeline/src/types.ts:17]:
```suggestion
} & HTMLPasteProps<"ol">;
```

File: packages/paste-core/components/timeline/src/types.ts
Comment (nkrantz) [inline packages/paste-core/components/timeline/src/types.ts:71]:
```suggestion
} & HTMLPasteProps<"li">;
```

File: packages/paste-core/components/timeline/__tests__/index.spec.tsx
Comment (nkrantz) [inline packages/paste-core/components/timeline/__tests__/index.spec.tsx:126]: suggestion(blocking): let's add TimelineItemIcon to these tests and any other elements we attach the data-paste-element attribute to

File: packages/paste-core/components/timeline/src/TimelineItem.tsx
Comment (nkrantz) [inline packages/paste-core/components/timeline/src/TimelineItem.tsx:56]:
```suggestion
            <TimelineItemIcon icon={icon} element=`{element}_ICON` />
```
Since we're not exposing TimelineItemIcon in the API , let's pass element down through the TimelineItem

File: packages/paste-core/components/timeline/src/TimelineItem.tsx
Comment (nkrantz) [inline packages/paste-core/components/timeline/src/TimelineItem.tsx:10]:
```suggestion
const TimelineItem = React.forwardRef<HTMLLIElement, TimelineItemProps>(
```

File: packages/paste-core/components/timeline/src/Timeline.tsx
Comment (nkrantz) [inline packages/paste-core/components/timeline/src/Timeline.tsx:51]:
```suggestion
const Timeline = React.forwardRef<HTMLOLElement, TimelineProps>(
```

File: packages/paste-core/components/timeline/src/TimelineItemGroup.tsx
Comment (nkrantz) [inline packages/paste-core/components/timeline/src/TimelineItemGroup.tsx:10]: suggestion(blocking): create TimelineItemGroupProps so that we can export it from the package. we like to export the types of each component we export as part of the API. that's also how the API docs are created

optional: move to types.ts file

File: packages/paste-core/components/timeline/src/TimelineItemGroup.tsx
Comment (nkrantz) [inline packages/paste-core/components/timeline/src/TimelineItemGroup.tsx:9]: suggestion(blocking): create TimelineItemGroupProps so that we can export it from the package. we like to export the types of each component we export as part of the API. that's also how the API docs are created

also use `HTMLPasteProps("li")` and `HTMLLIElement` for any components that are rendered as `li`s

optional: move to types.ts file

File: packages/paste-core/components/timeline/src/TimelineItemGroup.tsx
Comment (nkrantz) [inline packages/paste-core/components/timeline/src/TimelineItemGroup.tsx:52]: suggestion: let's call this and the Box on line 59 two different things (element_CONTENT)

File: packages/paste-core/components/timeline/src/types.ts
Comment (nkrantz) [inline packages/paste-core/components/timeline/src/types.ts:6]: suggestion: remove

**Review by nkrantz — COMMENTED** [review]
SO GOOD, sorry for so many comments 😄

File: packages/paste-core/components/timeline/__tests__/index.spec.tsx
Comment (PixeledCode) [reply packages/paste-core/components/timeline/__tests__/index.spec.tsx:71]: for some reason it didnt work. It works for first test, and then fails for all others if I do this

File: packages/paste-core/components/timeline/src/Timeline.tsx
Comment (PixeledCode) [reply packages/paste-core/components/timeline/src/Timeline.tsx:58]: its used for horizontal layout but if we are removing horizontal, might as well remove this

File: packages/paste-core/components/timeline/src/TimelineItemGroup.tsx
Comment (nkrantz) [inline packages/paste-core/components/timeline/src/TimelineItemGroup.tsx:9]:
```suggestion
export const TimelineItemGroup = React.forwardRef<HTMLLIElement, TimelineItemGroupProps>(
```

File: packages/paste-core/components/timeline/src/TimelineItemIcon.tsx
Comment (nkrantz) [reply packages/paste-core/components/timeline/src/TimelineItemIcon.tsx:13]: +1, what's the error here

File: packages/paste-core/components/timeline/src/index.tsx
Comment (nkrantz) [inline packages/paste-core/components/timeline/src/index.tsx:3]:
```suggestion

```

File: packages/paste-core/components/timeline/src/index.tsx
Comment (nkrantz) [inline packages/paste-core/components/timeline/src/index.tsx:5]:
```suggestion
export { TimelineItemGroup } from "./TimelineItemGroup";
export type { TimelineProps, TimelineItemProps, TimelineItemGroupProps } from "./types";
```

File: packages/paste-core/components/timeline/src/types.ts
Comment (nkrantz) [inline packages/paste-core/components/timeline/src/types.ts:57]:
```suggestion
   * @default null
```

File: packages/paste-core/components/timeline/src/types.ts
Comment (nkrantz) [inline packages/paste-core/components/timeline/src/types.ts:71]: (suggestion): since we don't need to export TimelineItemIcon as it's not a part of the API, you can remove all the comments from this type. we only use them for creating the API page of the docs site, but this type won't show up there.

File: packages/paste-core/components/timeline/src/types.ts
Comment (nkrantz) [inline packages/paste-core/components/timeline/src/types.ts:99]: praise: nice typings

File: packages/paste-core/components/timeline/src/types.ts
Comment (PixeledCode) [reply packages/paste-core/components/timeline/src/types.ts:57]: using `undefined` instead of null here

File: packages/paste-core/components/timeline/package.json
Comment (nkrantz) [inline packages/paste-core/components/timeline/package.json:41]: question: was skeleton loader added for use in the story file? I know eslint throws an error, but we don't add dependencies for use in the stories, just for use in the actual components we're exporting from the package. if it's annoying to see the red lines you can ignore the extraneous dependencies rule for the story file

**Review by roxanagomez — DISMISSED** [review]
Great job looks very good!

File: packages/paste-core/components/timeline/src/Timeline.tsx
Comment (krisantrobus) [reply packages/paste-core/components/timeline/src/Timeline.tsx:14]: I see this is still here. Are we removing? Just need to know before approval

**Review by krisantrobus — APPROVED** [review]
Great work! 🚀

**Review by roxanagomez — APPROVED** [review]

File: packages/paste-core/components/timeline/src/TimelineItemGroup.tsx
Comment (serifluous) [reply packages/paste-core/components/timeline/src/TimelineItemGroup.tsx]: true, ah well, maybe I'll care more or have a stronger opinion if it comes up again lol.

**Pull request merged**
